### PR TITLE
Add passphrase to ledger signing api

### PR DESCRIPTION
**Pull request by Yasuke**
This addresses #168; Instead of having the passphrase for signing
hardcoded to "Ledger.Crypto PassPhrase", allow the passphrase to be
provided. This way we aren't prevented from using private keys
created with tools like cardano-address and cardano-cli

### CODE REVIEW

**Review by silky — CHANGES_REQUESTED** [review]
Looks pretty good @Yasuke , maybe I could suggest a few potential refactorings/additional items:

1. Perhaps introduce a new function, `signMessage`, signObservation'`, that use the default, and the un-ticked ones take the passphrase. This way we don't have random empty-strings in a few places,
2. Might not hurt to wrap up the PassPhrase in a newtype, so that it's a bit more clear it's purpose (and to allow later for nuanced handling, if we wish it),
3. Maybe we could take this opportunity to have a few tests that verify this behaviour? (Not sure what tests we have already around verifying signatures and passphrases?)

Thanks for the work so far! :)

**Comment by volodyad** [comment]
Any progress on this?

**Comment by volodyad** [comment]
Tried to build from PR branch - and it fails 
src/Cardano/Wallet/Mock/Handlers.hs:128:23: error:
    • Couldn't match expected type ‘Eff effs MockWallet’
                  with actual type ‘BS.ByteString -> Eff effs0 MockWallet’
    • Probable cause: ‘newWallet’ is applied to too few arguments

**Comment by gege251** [comment]
@silky @Yasuke sorry for urging you, but this issue is a blocker for other work, could we prioritize merging this once it builds?

**Comment by silky** [comment]
Hi @gege251 ; sorry to hear it is blocking you. As you can see, we are working on it, and it will be merged in due course :)

**Comment by Yasuke** [comment]
Thank you for the feedback and review @silky! I have make Passphrase a distinct type, added prime functions that don't require a password, added some tests (and instances needed to make them work) for signing in both the ledger and oracle.

**Comment by Yasuke** [comment]
I also generalized the ledger signing function to work with any ByteArrayAccess instead of just TxIds

File: plutus-ledger/src/Ledger/CardanoWallet.hs [hunk 1]
```diff
@@ -68,10 +69,10 @@ newtype WalletNumber = WalletNumber { getWallet :: Integer }
     deriving anyclass (FromJSON, ToJSON)
 
 fromWalletNumber :: WalletNumber -> MockWallet
-fromWalletNumber (WalletNumber i) = fromSeed (BSL.toStrict $ serialise i)
+fromWalletNumber (WalletNumber i) = fromSeed (BSL.toStrict $ serialise i) ""
```
Comment (sjoerdvisscher) [inline plutus-ledger/src/Ledger/CardanoWallet.hs:72]: Use `noPassphrase` here.

File: plutus-use-cases/test/Spec/Future.hs [hunk 2]
```diff
@@ -203,7 +203,7 @@ settleEarly hdl = do
     void $ Trace.waitNSlots 1
 
 mkSignedMessage :: POSIXTime -> Value -> SignedMessage (Observation Value)
-mkSignedMessage time vl = Oracle.signObservation time vl (fst oracleKeys)
+mkSignedMessage time vl = Oracle.signObservation time vl (fst oracleKeys) ""
```
Comment (sjoerdvisscher) [inline plutus-use-cases/test/Spec/Future.hs:206]:
```suggestion
mkSignedMessage time vl = Oracle.signObservation time vl (fst oracleKeys) noPassphrase
```

File: plutus-use-cases/test/Spec/Stablecoin.hs [hunk 3]
```diff
@@ -63,7 +63,7 @@ coin = Stablecoin
     }
 
 signConversionRate :: POSIXTime -> ConversionRate -> SignedMessage (Observation ConversionRate)
-signConversionRate startTime rate = signObservation startTime rate oraclePrivateKey
+signConversionRate startTime rate = signObservation startTime rate oraclePrivateKey ""
```
Comment (sjoerdvisscher) [inline plutus-use-cases/test/Spec/Stablecoin.hs:66]:
```suggestion
signConversionRate startTime rate = signObservation startTime rate oraclePrivateKey noPassphrase
```

**Review by sjoerdvisscher — APPROVED** [review]
Really nice work!

File: plutus-ledger/src/Ledger/CardanoWallet.hs (hunk 1 above)
Comment (Yasuke) [reply plutus-ledger/src/Ledger/CardanoWallet.hs]: Adjusted, I used the ' version of the function that has no passphrase argument

File: plutus-use-cases/test/Spec/Stablecoin.hs (hunk 3 above)
Comment (Yasuke) [reply plutus-use-cases/test/Spec/Stablecoin.hs]: Also here, signObservation'

File: plutus-use-cases/test/Spec/Future.hs (hunk 2 above)
Comment (Yasuke) [reply plutus-use-cases/test/Spec/Future.hs]: Same with this one, I used signObservation' removing the passphrase argument

**Comment by Yasuke** [comment]
Pending CI, required adjustments have been made, thanks @sjoerdvisscher for the review

File: plutus-contract/src/Wallet/Emulator/Wallet.hs [hunk 4]
```diff
@@ -567,14 +567,14 @@ signTxWithPrivateKey :: (Member (Error WAPI.WalletAPIError) r) => PrivateKey ->
...
-    then pure (Ledger.addSignature pk tx)
+    then pure (Ledger.addSignature' pk tx)
     else throwError (WAPI.PrivateKeyNotFound pubK)
 
 -- | Sign the transaction with the given private keys,
 --   ignoring the list of public keys that the 'SigningProcess' is passed.
 signPrivateKeys :: [PrivateKey] -> SigningProcess
 signPrivateKeys signingKeys = SigningProcess $ \_ tx ->
-    pure (foldr Ledger.addSignature tx signingKeys)
+    pure (foldr (\pk ctx -> Ledger.addSignature' pk ctx) tx signingKeys)
```
Comment (silky) [inline plutus-contract/src/Wallet/Emulator/Wallet.hs:577]:
```suggestion
    pure (foldr Ledger.addSignature' tx signingKeys)
```

**Review by silky — APPROVED** [review]

File: plutus-contract/src/Wallet/Emulator/Wallet.hs (hunk 4 above)
Comment (Yasuke) [reply plutus-contract/src/Wallet/Emulator/Wallet.hs]: Good catch! fixed @silky

**Comment by Yasuke** [comment]
Rebased on current main

**Comment by volodyad** [comment]
hello, when this PR are planned to be merged?

**Comment by sjoerdvisscher** [comment]
Thanks for the reminder!

**Pull request merged**